docs/stdlib.md

```diff
@@ -0,0 +1,1554 @@
+
+::: Advanced
+
+In jaq, `walk(f)` is defined as `.. |= f`, whereas
+in `jq`, a definition similar to the following is used:
+
+```
+def walk(f): def rec: (.[]? |= rec) | f; rec;
+```
+
+This is a more efficient version of:
+
+```
+def walk(f): (.[]? |= walk(f)) | f;
+```
+
+We can show that in jaq, `.. |= f` and `jq`'s definition of `walk(f)` are equivalent.
+First, let us recall that `.. |= f` is equivalent to the following in jaq:
+
+```
+def rec_up: (.[]? | rec_up), .; rec_up |= f
+```
+
+We can thus unfold `.. |= f`:
+
+```
+..                   |= f  === (unfolding .. |= f)
+rec_up               |= f  === (unfolding rec_up)
+((.[]? | rec_up), .) |= f  === (because (l, r) |= f  ===  (l |= f) | (r |= f))
+((.[]? | rec_up) |= f)  | (. |= f)  === (because . |= f  ===  f)
+((.[]? | rec_up) |= f)  | f         === (because (l | r) |= f  ===  l |= (r |= f))
+(.[]? |= (rec_up |= f)) | f         === (because rec_up |= f  ===  .. |= f)
+(.[]? |= (.. |= f))     | f
+```
+
+We can see thus that
+`.. |= f` is equivalent to
+`(.[]? |= (.. |= f)) | f`.
+In the same sense,
+`walk(f)` is equivalent to
+`(.[]? |= walk(f)) | f`.
+We can conclude that `.. |= f` is equivalent to `walk(f)`.
+
+Note, however, that this equivalence does *not* hold in `jq`,
+because `jq`'s updates work differently than jaq's.
+The difference shows in particular when `f` returns multiple values.
+
+:::
+
+## Serialisation & Deserialisation
+
+### `tostring`
+
+The filter `tostring` converts its input to a string.
+Its output depends on the type of its input:
+
+- Text strings are returned unchanged, i.e. `"Hi" | tostring --> "Hi"`
+- Byte strings are interpreted as text string, i.e. `"Hi" | tobytes | tostring --> "Hi"`
+- Any other value is formatted compactly as if output by `jaq -c`. For example:
+
+  - `null | tostring --> "null"`,
+  - `[0, 1] | tostring --> "[0,1]"`,
+  - `{a: 1} | tostring --> "{\"a\":1}"`.
+
+::: Advanced
+
+String interpolation without an explicit format, such as
+`"\(null) and \([0, 1])" --> "null and [0,1]"`, behaves as if
+the output of every interpolated filter was piped through `tostring`.
+
+:::
+
+### `tonumber`
+
+The filter `tostring` takes as input either a number or a string.
+If the input is a number, it is returned unchanged;
+if the input is a string, it is parsed to a number, failing if this does not succeed.
+For example:
+
+- `  42   | tonumber --> 42`
+- ` "42"  | tonumber --> 42`
+- `"[42]" | try tonumber catch "fail" --> "fail"`
+
+### `toboolean`
+
+The filter `toboolean` is for booleans what `tonumber` is for numbers. 
+For example:
+
+- `  true   | toboolean --> true`
+- ` "true"  | toboolean --> true`
+- `"[true]" | try toboolean catch "fail" --> "fail"`
+
+### `fromjson`, `tojson`
+
+The filter `fromjson` takes a string as input,
+parses it to JSON values and yields them.
+For example:
+
+```
+"null true 0 \"foo\" [1] {\"foo\": 1}" | fromjson -->
+ null true 0  "foo"  [1] { "foo" : 1}
```

Comment:
Is two spaces between some of the outputs intentional? it felt inconsistent with the other examples :)